### PR TITLE
Batch assets creation

### PR DESCRIPTION
**Pull request by jorgemd24**
### Changes proposed in this Pull Request:

Closes part of https://github.com/woocommerce/google-listings-and-ads/issues/1780

While I was testing with different types of images and Google API limitations, I found that when uploading high-resolution images ( >3MB) the Ads API could return the following error:

``` The data written is too large. Split the request into smaller requests ```

Theoretically, the [Ads Library allows until 64MB payload](https://developers.google.com/google-ads/api/docs/best-practices/quotas#grpc_limitations) when using [gRPC](https://grpc.io/) protocol, in our case we use `rest` so we should not have this limitation however when I was testing this limitation I found that when the body is bigger than 30MB, the Ads API returns the error mention earlier. 

This PR allows uploading multiple images by creating asset batches, the challenge was to be able to maintain the atomicity for the asset group transaction. For that, I am creating first the assets and then I am mutating all operations related to the asset group assets (link and unlink assets with the asset group) so we keep the atomicity of the asset group operations.

If an asset with the same content is created multiple times, the Ads API will return the same ARN, which means, that the Ads API does not create duplicated assets. 

### Detailed test instructions:

1. The WooCommerce Connect Server by [default](https://hapi.dev/api/?v=21.1.0), allows 1MB maximum as a payload so you will need to check this branch: `add/max-payload-size-google-proxy`  that allows bigger payloads and run WCS locally. 
2. Create an empty asset group `POST /gla/ads/campaigns/asset-groups` with the following body: ` { "campaign_id" : YOUR_CAMPAIGN_ID }`
2. Make a call to `PUT /gla/ads/campaigns/asset-groups/YOUR_ASSET_GROUP_ID` with the below body.
<details>
<summary> Click here. The body contains high-resolution images (>3MB , maximum is 5MB each image) and the maximum number of assets for each type. (worst case scenario) </summary>

```
{
   "assets": [
       {
           "id" : null,
           "field_type": "headline",
           "content": "headline 1"
       },
       {
           "id" : null,
           "field_type": "headline",
           "content": "headline 2"
       },
       {
           "id" : null,
           "field_type": "headline",
           "content": "headline 3"
       },
       {
           "id" : null,
           "field_type": "headline",
           "content": "headline 4"
       },
       {
           "id" : null,
           "field_type": "headline",
           "content": "headline 5"
       },
       {
           "id" : null,
           "field_type": "long_headline",
           "content": "long headline 1"
       },
       {
           "id" : null,
           "field_type": "long_headline",
           "content": "long headline 2"
       },
       {
           "id" : null,
           "field_type": "long_headline",
           "content": "long headline 3"
       },
       {
           "id" : null,
           "field_type": "long_headline",
           "content": "long headline 4"
       },
       {
           "id" : null,
           "field_type": "long_headline",
           "content": "long headline 5"
       },
       {
           "id" : null,
           "field_type": "description",
           "content": "description 1"
       },
       {
           "id" : null,
           "field_type": "description",
           "content": "description 2"
       },
       {
           "id" : null,
           "field_type": "description",
           "content": "description 3"
       },
       {
           "id" : null,
           "field_type": "description",
           "content": "description 4"
       },
       {
           "id" : null,
           "field_type": "description",
           "content": "description 5"
       },
       {
           "id" : null,
           "field_type": "business_name",
           "content": "business name test"
       },       
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_1.jpg"
       },
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_2.png"
       },
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_3.png"
       },
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_4.png"
       },
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_5.png"
       },
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_6.png"
       },
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_7.png"
       },
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_8.png"
       },
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_9.png"
       },
       {
           "id" : null,
           "field_type": "square_marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_square_10.png"
       },   
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_1.jpg"
       },
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_2.png"
       }, 
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_3.png"
       }, 
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_4.png"
       },
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_5.png"
       },
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_6.png"
       },
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_7.png"
       },
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_8.png"
       },
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_9.png"
       },
       {
           "id" : null,
           "field_type": "marketing_image",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_marketing_10.png"
       },
       {
           "id" : null,
           "field_type": "logo",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_logo_1.png"
       },
       {
           "id" : null,
           "field_type": "logo",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_logo_2.jpeg"
       },
       {
           "id" : null,
           "field_type": "logo",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_logo_3.jpeg"
       },
       {
           "id" : null,
           "field_type": "logo",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_logo_4.jpeg"
       },
       {
           "id" : null,
           "field_type": "logo",
           "content": "https://stormy-rabbit.jurassic.ninja/wp-content/uploads/2023/01/test_logo_5.jpg"
       }
   ]
}
```
</details>

4. You should see that the asset group has been edited successfully. It can take some time depending on your connection and device. The total body is about 100MB.

### Additional details:
- As we are using some new imports from the GoogleAds Library, you will need to remove your vendor folder `rm -rf vendor/* ` and install again the dependencies `composer install`.
- You may need to increase your memory limit so your device can process all images. You can add the following line in your `wp-config` file to increase the memory limit: `define( 'WP_MEMORY_LIMIT', '512M' );`
- The maximum number of marketing images is 20 and it is the sum of marketing images + square_images + portrait marketing images.
- The maximum is 5 logo images.

### CODE REVIEW

File: src/API/Google/AdsAsset.php [hunk 1]
```diff
@@ -38,13 +39,36 @@ class AdsAsset implements OptionsAwareInterface {
...
+	 * @var int
+	 */
+	protected const MAX_PAYLOAD_BYTES = 30 * 1024 * 1024;
+
+	/**
+	 * Maximum image size in bytes.
+	 *
+	 * @var int
+	 */
+	protected const MAX_IMAGE_SIZE = 5 * 1024 * 1024;
```
Comment (puntope) [inline src/API/Google/AdsAsset.php]: 💅  What about adding here `_BYTES` suffix as we did in `MAX_PAYLOAD_BYTES`?

File: src/API/Google/AdsAsset.php [hunk 2]
```diff
@@ -38,13 +39,36 @@ class AdsAsset implements OptionsAwareInterface {
 	 */
 	protected WP $wp;
 
+	/**
+	 * The Google Ads Client.
+	 *
+	 * @var GoogleAdsClient
+	 */
+	protected $client;
```
Comment (puntope) [inline src/API/Google/AdsAsset.php]: 💅  We can add here also the type GoogleAdsClient

File: src/API/Google/AdsAsset.php [hunk 3]
```diff
@@ -93,16 +117,109 @@ protected function get_asset_type_by_field_type( string $field_type ): int {
...
+	 * @param string $url The image url.
+	 *
+	 * @return array The image data.
+	 * @throws Exception If the image url is not a valid url or the image size is too large.
+	 */
+	protected function get_image_data( string $url ): array {
+		$image_data = $this->wp->wp_remote_get( $url );
+
+		if ( is_wp_error( $image_data ) || empty( $image_data['body'] ) ) {
+			throw new Exception( 'Incorrect image asset url.' );
```
Comment (puntope) [inline src/API/Google/AdsAsset.php]: The first time I tested I was getting Curl remote error timeouts here. Which was not really an  `Incorrect image asset url`. After debugging I saw the CURL error but I was thinking it was a problem in the URLs. 

I would change this Exception to other thing (maybe getting the error?) or at least including the url could help more.

File: src/API/Google/AdsAsset.php [hunk 4]
```diff
@@ -93,16 +117,109 @@ protected function get_asset_type_by_field_type( string $field_type ): int {
...
+	}
+
+	/**
+	 * Returns a list of batches of assets.
+	 *
+	 * @param array $assets A list of assets.
+	 * @param int   $max_size The maximum size of the payload in bytes.
+	 *
+	 * @return array A list of batches of assets.
+	 * @throws Exception If the image url is not a valid url or if the field type is not supported.
```
Comment (puntope) [inline src/API/Google/AdsAsset.php]: This in theory also triggers exception when the size is over the maximum image size.

File: src/API/Google/AdsAsset.php [hunk 5]
```diff
@@ -175,4 +286,29 @@ public function convert_asset( GoogleAdsRow $row ): array {
...
 		];
 	}
+
+	/**
+	 * Send a batch of operations to mutate assets.
+	 *
+	 * @param MutateOperation[] $operations
+	 *
+	 * @return array A list of Asset's ARN created.
+	 * @throws ApiException If any of the operations fail.
```
Comment (puntope) [inline src/API/Google/AdsAsset.php:297]: Maybe we can import `Google\ApiCore\ApiException` allowing to navigate over ApiException in the IDE.

**Review by puntope — APPROVED** [review]
✅  LGTM

I had troubles by my own environment timeout limitation. But finally I was able to test with the large request.

- I tested with the old WCS version and got Request size error.
- I tested with the patched WCS version running locally and was successful.
- Getting the assets afterwards shows the right images.
- Checked the code. Just minor 💅  suggestions

Approved in advance

**Comment by jorgemd24** [comment]
Thanks @puntope for your time and for spending so much time trying to figure out the problems with your timeout limitations! It is really appreciated 🙏

**Pull request merged**